src/server/generic_family_test.cc

```diff
@@ -412,6 +412,49 @@ TEST_F(GenericFamilyTest, Scan) {
   EXPECT_EQ(resp, "");
 }
 
+TEST_F(GenericFamilyTest, ScanWithAttr) {
+  Run({"flushdb"});
+
+  Run({"set", "hello", "world"});
+  Run({"set", "foo", "bar"});
+
+  Run({"expire", "hello", "1000"});
+
+  auto resp = Run({"scan", "0", "attr", "v"});
+  auto vec = StrArray(resp.GetVec()[1]);
+  EXPECT_EQ(1, vec.size());
+  EXPECT_EQ(vec[0], "hello");
+
+  resp = Run({"scan", "0", "attr", "p"});
+  vec = StrArray(resp.GetVec()[1]);
+  EXPECT_EQ(1, vec.size());
+  EXPECT_EQ(vec[0], "foo");
+
+  // before run get "foo", scan with a attr should return empty set
+  resp = Run({"scan", "0", "attr", "a"});
+  vec = StrArray(resp.GetVec()[1]);
+  EXPECT_EQ(0, vec.size());
+
+  // before run get "foo", scan with a attr should return "foo" and "hello"
+  resp = Run({"scan", "0", "attr", "u"});
+  vec = StrArray(resp.GetVec()[1]);
+  EXPECT_EQ(2, vec.size());
+
+  ASSERT_THAT(Run({"get", "foo"}), "bar");
```

Comment:
@romange  In this test case, i found that running "get" of some key, it wiil not make the key `touched`, so this test case failed, can you have a look what is wrong.

Reply:
it makes the key touched and you can verify it by adding debug printings for example at line db_slice.cc:624

Reply:
> it makes the key touched and you can verify it by adding debug printings for example at line db_slice.cc:624

i see what is wrong, HasExpire() works on it->second object(which is value), while HasTouched() works on it->first object(which is key), is that expected?

Reply:
yes, this by design though I agree it's confusing.
the reason is that if you call "SADD foo bar"   and then `SET foo bar`  then all the attributes on the value (second) will be reset. if we want to preserve the attributes we should use keys (first)

Reply:
now this pr is ready for review :)